include/deal.II/lac/sparse_matrix.templates.h

```diff
@@ -1328,6 +1352,8 @@ SparseMatrix<number>::precondition_Jacobi (Vector<somenumber>       &dst,
   AssertDimension (dst.size(), n());
   AssertDimension (src.size(), n());
 
+  check_nonzero_diagonal(*this);
```

Comment:
Does the optimizer remove the loop over all rows?

Personally, I'd just guard the call to <code>check_nonzero_diagonal</code> with <code>#ifdef DEBUG</code> in order to not rely on any optimization.